msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
+        if gene_conversion_rate is None:
+            # It doesn't matter what the track_length is, just set a
+            # value to keep the low-level code happy.
+            gene_conversion_track_length = 1
         else:
-            if sequence_length is None:
+            raise ValueError(
+                "Must specify track length when simulating gene conversion"
+            )
+    else:
+        if gene_conversion_rate is None:
+            raise ValueError(
+                "Must specify gene conversion rate along with track length"
+            )
+        gene_conversion_track_length = float(gene_conversion_track_length)
+
+    # TODO we many need to have different ploidy values for the different models.
+    # For example, dtwf is an inherently diploid process and anything else doesn't
+    # make sense. However, having different defaults for ploidy will lead to
+    # confusing problems when we have mixed models, so perhaps we should raise
+    # an error from models that require ploidy=2 and force them to explicitly
+    # set it.
+    # TODO Might also be worth considering setting ploidy=2 by default. This
+    # *is* likely to be the most common application, and those doing haploid
+    # stuff can set ploidy=1, if they like.
+    ploidy = 1 if ploidy is None else ploidy
+    if not core.isinteger(ploidy):
+        raise TypeError("ploidy must be an integer")
+    ploidy = int(ploidy)
+    if ploidy < 1:
+        raise ValueError("ploidy must be >= 1")
+
+    model, model_change_events = _parse_model_arg(model)
+    is_dtwf = isinstance(model, DiscreteTimeWrightFisher)
+
+    # Check the demography. If no demography is specified, we default to a
+    # single-population model with a given population size.
+    if demography is None:
+        if is_dtwf:
+            # A default size of 1 isn't so smart for DTWF and almost certainly
+            # an error.
+            if population_size is None:
                 raise ValueError(
-                    "Must specify a sequence_length for simulations of a "
-                    "genome with recombination or gene conversion without "
-                    "rate map"
+                    "When using the DTWF model, the population size must be set "
+                    "explicity, either using the population_size or demography "
+                    "arguments."
                 )
+        population_size = 1 if population_size is None else float(population_size)
+        demography = demog.Demography.simple_model(population_size)
+    elif isinstance(demography, demog.Demography):
+        if population_size is not None:
+            raise ValueError("Cannot specify demography and population size")
+    else:
+        raise TypeError("demography argument must be an instance of msprime.Demography")
+    # TODO should we validate that the population sizes are positive integers
+    # for DTWF? It's not actually required, but could help catch some errors?
```

Comment:
Is there a reason to not do this for an explicitly finite population model?

Reply:
Mainly because it gets caught at runtime during the simulation - we'll get a "population size has gone to zero" (or similar) error. It's slightly murky, because DTWF rounds the population sizes because of exponential growth, so we do actually handle non-integer sizes within the simulation. It seems a bit odd to say you must have integer sizes on the way in, then, and could end up being awkward. :shrug: I don't know though.

Reply:
Good point about needing to round anyway. In that case, I think it feels cleaner to not bother with a check here either.